### ISSUE TITLE
No biomaterial libraries express this feature, despite having biomaterials

### ISSUE THREAD

**Issue by mboudet** [tripal/tripal_analysis_expression#224]
Hello,

I currently have an issue where the "Expression" tab in features show "No biomaterial libraries express this feature", despite having two biomaterials expressing it.

The issue seems to stem from the $has_exp value not being set to TRUE.

In the 'tripal_biomaterial.theme.inc' file, there is a loop over all expressions after an sql call. Afterward, there is a check on values.

`    if($first_sig == 1) {`
`      $last_signal = $exp->signal;`
`      $first_sig = 0;`
`    }`
`    if ($exp->signal != $last_signal and $exp->signal != 0) {`
`      $has_exp = TRUE;`
`    }`

The "$has_exp" variable is set to TRUE if the current expression value is different from the previous expression value, and not 0.

There are several cases where this lead to the expression tab being empty : 

- Only one expression is in DB 
- Several expression with the same value on all expressions.
- One expression with a non zero value, and one expression with a 0 value.

Is there any reason for this check? (removing some duplicata maybe?)
`$exp->signal != $last_signal`

(These are edge cases for sure, but it was easier for testing to upload the same expression file..)

**Comment by bradfordcondon** [comment]
hello, thanks for reporting the issue!

I agree this whole loop is really strange.  it actually looks like the only expression value that matters is the last one in the query, since that is what will set `$has_exp`.   I I also dont see the purpose of checking the current signal against the last signal.

What happens if you just get rid of all that nonsense?

IE something like below- basically just set has_exp to false if there are no results from the query.  I dont see why you wouldnt want to display a biomaterial set with all 0 results, thats still data...

```php

  $has_exp = TRUE;

  $feature_id = chado_get_id_from_nid('feature', $node->nid);

  $sql = "SELECT ER.signal, B.name, B.biomaterial_id, B.description FROM {elementresult} ER
           INNER JOIN {element} E ON E.element_id = ER.element_id
           INNER JOIN {feature} F ON F.feature_id = E.feature_id
           INNER JOIN {quantification} Q ON Q.quantification_id = ER.quantification_id
           INNER JOIN {acquisition} AQ ON AQ.acquisition_id = Q.acquisition_id
           INNER JOIN {assay} A ON A.assay_id = AQ.assay_id
           INNER JOIN {assay_biomaterial} AB ON AB.assay_id = A.assay_id
           INNER JOIN {biomaterial} B ON B.biomaterial_id = AB.biomaterial_id
         WHERE
           F.feature_id = :feature_id";

  $expressions = chado_query($sql, [":feature_id" => $feature_id]);

  $json_exp = "[";

  $num_matches = $expressions->rowCount();

  if ($num_matches == 0){
    $has_exp = FALSE;

  }

  foreach ($expressions as $exp) {
    $tis = chado_get_property([
      'table' => 'biomaterial',
      'id' => $exp->biomaterial_id,
    ], ['type_name' => 'tissue', 'cv_name' => 'biomaterial_property']);

    if ($tis) {
      $tissue = $tis->value;
    }
    else {
      $tissue = '';
    }
```

**Comment by mboudet** [comment]
Agreed, the check for a value of 0 is also weird.

Removing this part and just setting $has_exp depending on the query results, does not seems to cause any issues in my tripal instance, but my test case is limited.

**Comment by bradfordcondon** [comment]
I dont have a tripal 2 instance handy (we dont officially support tripal 2 on any of our modules any more, we dont have the ability to support both branches). 

It would be really great if you could check the fix on say a development site with real data so we could make a PR.  If not I understand- thanks again for pointing it out!

**Comment by mboudet** [comment]
I should be able to test it on a developpement site soon.

However, it seems that this check exist in the tripal 3 version too (in tripal_biomaterial.theme.inc
 and tripal_protocol.theme.inc), though I am not sure if the bug appears.

**Comment by bradfordcondon** [comment]
yes I'm actually working on this module today and noticed that.  The short story is I am the one who converted the module from Tripal 2 to Tripal 3, but its a very large project.  So, there is code that could be cut that I missed.

I believe that this whole file is unused in Tripal 3, as the functionality is provided by the **field**.

**Comment by mboudet** [comment]
Alright, I will make the change on a dev site, and report back if there are not other issues.

**Comment by mboudet** [comment]
So, after more testing : 
Removing the 'if ($exp->signal != $last_signal)' shows no issues, both the heatmap and the expression panel work properly.

However, removing the  if($exp->signal != 0) cause some issues when all expression values equal 0.
In expression.js, we use maxHeat (the maximum expression value) to define the size of the graph. Namely 
`((graphHeight*d.intensity)/maxHeat))`, which is a division by 0.

It doesn't actually have an effect on the graph because it is empty anyway.

I think the best way to proceed would be : 

- Remove the ''if ($exp->signal != $last_signal)' check
- Keep the 0 signal check
- Change the "No biomaterial libraries express this feature" message to something a little clearer, to indicate that while some biomaterials are linked to this feature, none have an expression value above 0.

**Comment by mboudet** [comment]
Fixed by #251
